src/west/commands/project.py

```diff
@@ -212,38 +212,141 @@ def __init__(self):
             _wrap('''
             Updates all projects according to the manifest file, `west.yml`,
             in the manifest repository.
-            All projects will be checked out in detached HEAD and the local
-            manifest-rev branch will be reset hard to same HEAD.
 
-            Any local branches will not be altered by this command.
+            By default:
 
-            This command will not touch the manifest repository.
+            1. West itself is updated to the revision in the manifest
+            file. If revision is a branch, west will update to the tip
+            of the branch.
+
+            2. The local per-project manifest-rev branches (which are
+            managed exclusively by west) will be hard reset to the
+            updated commits fetched from each project's remote.
+
+            3. All projects will have detached HEADs checked out at
+            the manifest-rev commits. Locally checked out branches are
+            unaffected but will no longer be checked out by default.
+            West will print information on how to get back to those
+            branches or update them to the new manifest-rev versions.
+
+            You can avoid 1. using --no-update. You can (sometimes)
+            avoid 3. using --keep-descendants and/or --rebase.
+            See below for more information.
+
+            This command does not change the contents of the manifest
+            repository.
             '''))
 
     def do_add_parser(self, parser_adder):
-        return _add_parser(parser_adder, self, _no_update_arg,
+        return _add_parser(parser_adder, self,
+                           _no_update_arg,
+                           _arg('-k', '--keep-descendants',
+                                action='store_true',
+                                help=_wrap('''
+                                If a local branch is checked out and is a
+                                descendant commit of the new manifest-rev,
+                                then keep that descendant branch checked out
+                                instead of detaching HEAD. This takes priority
```

Comment:
Should we specify here also `and if the manifest revision is a SHA`. Since otherwise it will check out a branch. And what happens if you use `-k` and the manifest rev is a branch?

Reply:
nope, even if revision is a branch, it will still chackout as detached HEAD

Reply:
> And what happens if you use -k and the manifest rev is a branch?

It will still compare their relationships in the graph:

```
$ git log --oneline --decorate --graph tmp2~..tmp1
* 3e8fc3c0c1 (tmp1) modem: wncm14a2a: Convert to new DT_<COMPAT>_<INSTANCE> defines
* fe8edc2884 (tmp2) display: ssd1306: Convert to new DT_<COMPAT>_<INSTANCE> defines
$ git merge-base --is-ancestor tmp2 tmp1; echo $?
0
$ git merge-base --is-ancestor tmp1 tmp2; echo $?
1
```

Reply:
>nope, even if revision is a branch, it will still chackout as detached HEAD

Oh right, I guess I wasn't clear about that one then. Thanks for clarifying

Reply:
>It will still compare their relationships in the graph:

Right, so it will simply use the 2 `HEAD`s of each branch as commits to run `merge-base` on